ViTool/Models/TranslateXmlToTxTAlgorithm.cs

```diff
@@ -0,0 +1,165 @@
+            progress.Report(progressReportModel);
+        }
+
+        private async Task ProcessFilesParalelAsync(string directory, string xmlExt, List<string> classes, string[] files, IProgress<ProgressReportModel> progress)
+        {
+            int timeForOneRecord;
+            int procesorsCount = Environment.ProcessorCount;
+
+            if (procesorsCount >= 4)
+                procesorsCount -= 2;
+
+            await Task.Run(() => Parallel.ForEach<string>(files, new ParallelOptions { MaxDegreeOfParallelism = procesorsCount }, src =>
+            {
+                Stopwatch watch = new Stopwatch();
+                watch.Start();
+                ProgressReportModel progressReportModel = new ProgressReportModel();
+
+                if (Path.GetExtension(src) != xmlExt)
+                    return;
+
+                XmlDocument doc = new XmlDocument();
+                doc.Load(src);
+
+                int frameWidth = int.Parse(doc.DocumentElement.SelectSingleNode("/annotation/size/width").InnerText);
+                int frameHeight = int.Parse(doc.DocumentElement.SelectSingleNode("/annotation/size/height").InnerText);
+
+                XmlNodeList nodes = doc.DocumentElement.SelectNodes("/annotation/object");
+
+                List<TxtDefectRow> defects = CreateMirroredDefects(classes, doc, frameWidth, frameHeight, nodes);
+
+                if (defects.Count == 0)
+                    return;
+
+                string xmlFilename = Path.GetFileNameWithoutExtension(src) + ".txt";
+                string mirroredXmlSrc = Path.Combine(directory, xmlFilename);
+
+                SaveToTxt(defects, mirroredXmlSrc);
+                Output.Add(mirroredXmlSrc);
+
+                watch.Stop();
+
+                progressReportModel.FilesProcessed.Add(mirroredXmlSrc);
+                progressReportModel.TimeConsumedByProcessedFiles = (int)((HowMuchThereIs - Output.Count) * watch.Elapsed.TotalMilliseconds * 0.001);
+                progressReportModel.PercentageComplete = (Output.Count * 100) / HowMuchThereIs;
+                progress.Report(progressReportModel);
+            }));
+        }
+
+        private List<TxtDefectRow> CreateMirroredDefects(List<string> classes, XmlDocument doc, int frameWidth, int frameHeight, XmlNodeList nodes)
+        {
+            List<TxtDefectRow> defects = new List<TxtDefectRow>();
+            foreach (XmlNode currentNode in nodes)
+            {
+                TxtDefectRow defectRow = CreateMirroredDefect(classes, doc, frameWidth, frameHeight, currentNode);
+                if (defectRow != null)
+                    defects.Add(defectRow);
+            }
+
+            return defects;
+        }
+
+        private TxtDefectRow CreateMirroredDefect(List<string> classes, XmlDocument doc, int frameWidth, int frameHeight, XmlNode node)
+        {
+            TxtDefectRow defectRow = new TxtDefectRow();
+
+            int xmin = int.Parse(node.SelectSingleNode("bndbox/xmin").InnerText);
```

Comment:
i have seen this code few time, check if you can just add extension method to inner text type